### PR TITLE
Implemented filtering options for eth_subscribe with logs #27842

### PR DESCRIPTION
**Pull request by vadimstruts**
PR represents the implementation of advanced functionality for log subscription. It provides filtering options for retrieving logs

For example, the next JS code illustrates new approach:
```
window.ethereum.on('message', (x, y, z) => console.log('message event: ', x, y, z))
window.ethereum.send({ id:1, method: 'eth_subscribe', params:['logs', {"address":  ["0xdac17f958d2ee523a2206206994597c13d831ec7", "0x51fe8e8fa3f9f10288fa8c2aff1400f887d21c42"]}]}, console.log)
```
As you can see, only the logs for the next two addresses will be retrieved each time:
```
"0xdac17f958d2ee523a2206206994597c13d831ec7", 
"0x51fe8e8fa3f9f10288fa8c2aff1400f887d21c42"
```
Full list of supported filtering options described here: [https://www.quicknode.com/docs/ethereum/eth_getLogs](https://www.quicknode.com/docs/ethereum/eth_getLogs)

Resolves [brave/brave-browser/issues/27842](https://github.com/brave/brave-browser/issues/27842)

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues/27842) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Subscribe for logs with additional filtering options
```
window.ethereum.on('message', (x, y, z) => console.log('message event: ', x, y, z))
window.ethereum.send({ id:1, method: 'eth_subscribe', params:['logs', {"address":  ["0xdac17f958d2ee523a2206206994597c13d831ec7", "0x51fe8e8fa3f9f10288fa8c2aff1400f887d21c42"]}]}, console.log)
```
Response:
```
{
    "id": 1,
    "jsonrpc": "2.0",
    "result": "0xac2fa41a9975eeac8a2ef8f8ae1c97e8"
}
```
2. Receiving logs only for specified address (or another filtering options):  `0xdac17f958d2ee523a2206206994597c13d831ec7`

3.  Unsubscribe with using subscription id:

```
window.ethereum.on('message', (x, y, z) => console.log('message event: ', x, y, z))
window.ethereum.send({ id:5, method: 'eth_unsubscribe', params:['0xac2fa41a9975eeac8a2ef8f8ae1c97e8']}, console.log)
```
4. Logging is stopped

### CODE REVIEW

File: browser/brave_wallet/ethereum_provider_impl_unittest.cc [hunk 1]
```diff
@@ -2171,6 +2180,95 @@ TEST_F(EthereumProviderImplUnitTest, EthSubscribeLogs) {
...
+          EXPECT_EQ(*toBlock, "0xab65");
+
+          auto* topics = params->front().GetDict().FindList("topics");
+          EXPECT_TRUE(topics);
+          auto topic1 = std::find(topics->begin(), topics->end(), "0x2edc");
+          EXPECT_TRUE(topic1 != topics->end());
+          auto topic2 = std::find(topics->begin(), topics->end(), "0xb832");
+          EXPECT_TRUE(topic2 != topics->end());
+          auto topic3 = std::find(topics->begin(), topics->end(), "0x8dc8");
+          EXPECT_TRUE(topic3 != topics->end());
```
Comment (supermassive) [inline browser/brave_wallet/ethereum_provider_impl_unittest.cc:2225]: Worth just parsing hardcoded json string to base::Value to compare with payload

File: components/brave_wallet/common/eth_request_helper.cc [hunk 2]
```diff
@@ -677,19 +677,84 @@ bool ParseEthSendRawTransactionParams(const std::string& json,
...
+    const auto* vvalue_list = vvalue->GetIfList();
+
+    if (vvalue_list == nullptr)
+      return;
+
+    for (const auto& topics_list_item : (*vvalue_list)) {
+      if (const auto* topics_list_item_str = topics_list_item.GetIfString()) {
+        arr_out->emplace_back(*topics_list_item_str);
+      }
+    }
```
Comment (supermassive) [inline components/brave_wallet/common/eth_request_helper.cc:735]: I think topics filter for get_ethLogs and eth_newFilter has same format
https://ethereum.org/ru/developers/docs/apis/json-rpc/#eth_newfilter
which means [A, [B, C]] and [A, B, C] are different filters, but this parser flattens such array. Nulls are also supposed to be supported.

Comment (supermassive) [reply components/brave_wallet/common/eth_request_helper.cc:735]: Well it seems to me logging params should be transparent to browser, i.e. we should get filter params as base::Value from renderer and pass it to rpc endpoint as is.
So interface should look like this
```
void JsonRpcService::EthGetLogs(const std::string& chain_id,
                                const base::Value& params,
                                EthGetLogsCallback callback)
```

File: browser/brave_wallet/ethereum_provider_impl_unittest.cc (hunk 1 above)
Comment (vadimstruts) [reply browser/brave_wallet/ethereum_provider_impl_unittest.cc]: fixed

File: components/brave_wallet/common/eth_request_helper.cc (hunk 2 above)
Comment (vadimstruts) [reply components/brave_wallet/common/eth_request_helper.cc]: fixed

File: components/brave_wallet/browser/ethereum_provider_impl.cc [hunk 3]
```diff
@@ -1233,13 +1276,13 @@ void EthereumProviderImpl::CommonRequestOrSendAsync(base::ValueView input_value,
   } else if (method == kWeb3ClientVersion) {
     Web3ClientVersion(std::move(callback), std::move(id));
   } else if (method == kEthSubscribe) {
-    std::string event_type;
-    if (!ParseEthSubscribeParams(normalized_json_request, &event_type)) {
+    base::Value params;
+    if (!ParseEthSubscribeParams(normalized_json_request, &params)) {
```
Comment (supermassive) [inline components/brave_wallet/browser/ethereum_provider_impl.cc:1280]: `ParseEthSubscribeParams` should check if `params` is a list of one or two elements and return them. The first one should be a string, the second one should be a dict(if exists)

EthSubscribe gets them as args:
`EthereumProviderImpl::EthSubscribe(const std::string& event_type, absl::optional<base::Value::Dict> filter,...)` and I don't think we need to do any validation of filter's content(for example we still don't do full validation of address per spec *address (optional): either a string representing an address or an array of such strings.*) - let that be backend responsibility:

```
  } else if (event_type == kEthSubscribeLogs && filter) {
    const auto gen_res = generateHexBytes(eth_log_subscriptions_);

    if (std::get<0>(gen_res)) {
      eth_logs_tracker_.Start(base::Seconds(kLogTrackerDefaultTimeInSeconds));
    }

    eth_logs_tracker_.AddSubscriber(std::get<1>(gen_res),
                                    std::move(*filter));

    std::move(callback).Run(std::move(id), base::Value(std::get<1>(gen_res)),
                            false, "", false);
  } else {
```

File: components/brave_wallet/browser/json_rpc_service.cc [hunk 4]
```diff
@@ -2181,12 +2178,56 @@ void JsonRpcService::EthGetLogs(const std::string& chain_id,
...
+  base::Value::List contract_addresses, topics;
+  load_params(from_block, to_block, block_hash, &contract_addresses, &topics);
   auto internal_callback =
       base::BindOnce(&JsonRpcService::OnEthGetLogs,
                      weak_ptr_factory_.GetWeakPtr(), std::move(callback));
   RequestInternal(
       eth::eth_getLogs(from_block, to_block, std::move(contract_addresses),
-                       std::move(topics), ""),
+                       std::move(topics), block_hash),
       true, network_url, std::move(internal_callback));
```
Comment (supermassive) [inline components/brave_wallet/browser/json_rpc_service.cc:2231]: We don't need to look into params. Just change eth::eth_getLogs to something like:
```
std::string eth_getLogs(base::Value::Dict filter_options) {
  base::Value::List params;
  params.Append(std::move(filter_options));
  base::Value::Dict dictionary =
      GetJsonRpcDictionary("eth_getLogs", std::move(params));
  return GetJSON(dictionary);
}
```
If at some point we need a c++ way to create such filter then that should be 
```
base::Value::Dict MakeEthLogFilter(const std::string& from_block_quantity_tag,
                        const std::string& to_block_quantity_tag,
                        base::Value::List addresses,
                        base::Value::List topics,
                        const std::string& block_hash);
```
which makes a filter dict not a rpc call structure

File: components/brave_wallet/browser/ethereum_provider_impl.cc (hunk 3 above)
Comment (yrliou) [reply components/brave_wallet/browser/ethereum_provider_impl.cc:1280]: I agree with just letting RPC endpoint to validate/handle the filter keys.
Just a side note, it seems the RPC endpoint can support eth_subscribe logs without filter, so in additional to support filter, we probably want to keep what we currently support (without filter) too.

Comment (vadimstruts) [reply components/brave_wallet/browser/ethereum_provider_impl.cc]: fixed

File: components/brave_wallet/browser/json_rpc_service.cc (hunk 4 above)
Comment (vadimstruts) [reply components/brave_wallet/browser/json_rpc_service.cc]: fixed

**Review by supermassive — APPROVED** [review]
LGTM

**Pull request merged**